### PR TITLE
Add OptionsPremiumPricer Contract

### PR DESCRIPTION
**Pull request by chuddster**
…vault

### CODE REVIEW

File: contracts/core/OptionsPremiumPricer.sol [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    function getPremium(
+        uint256 st,
+        uint256 t,
+        bool isPut
+    ) external returns (uint256 premium) {
+        volatilityOracle.commit();
+
+        uint256 sp = priceOracle.latestAnswer().div(priceOracle.decimals());
+        uint256 v = volatilityOracle.vol();
+        uint256 rf = getAssetRiskFreeRate();
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:53]: todo: incorporate rf into black scholes pricing, unless we can ignore and just call [quoteAll](https://etherscan.io/address/0xCCdfCB72753CfD55C5afF5d98eA5f9C43be9659d#code) in keep3r volatility contract

File: contracts/core/OptionsPremiumPricer.sol [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    /**
+     * @notice Calculates the option's delta
+     * Formula reference: `d_1` in https://www.investopedia.com/terms/b/blackscholes.asp
+     * http://www.optiontradingpedia.com/options_delta.htm
+     * @param st is the strike price of the option
+     * @param t is the time to expiration in years
+     */
+    function getOptionDelta(uint256 st, uint256 t)
+        external
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:115]: same here for rf 

https://github.com/ribbon-finance/rvol/pull/3/files#r647943352

File: contracts/core/OptionsPremiumPricer.sol [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        uint256 sig = (((1e18 * sigma) / sigmaB) * t) / 365;
+
+        uint256 sSQRT = (v * Math.sqrt2((1e18 * t) / 365)) / 1e9;
+
+        uint256 d1 = (1e18 * Math.ln((Math.FIXED_1 * sp) / st)) / Math.FIXED_1;
+        d1 = ((d1 + sig) * 1e18) / sSQRT;
+
+        delta = Math.ncdf((Math.FIXED_1 * d1) / 1e18);
+    }
+
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:132]: need to run tests

File: contracts/core/OptionsPremiumPricer.sol [hunk 4]
```diff
@@ -0,0 +1,139 @@
+//SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.7.3;
+
+import {IVolatilityOracle} from "../interfaces/IVolatilityOracle.sol";
+import {IPriceOracle} from "../interfaces/IPriceOracle.sol";
+import {ICompToken} from "../interfaces/ICompToken.sol";
+import {IERC20Detailed} from "../interfaces/IERC20Detailed.sol";
+import {Math} from "../libraries/Math.sol";
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+import {Welford} from "../libraries/Welford.sol";
```
Comment (kenchangh) [inline contracts/core/OptionsPremiumPricer.sol:10]: Can remove this import

File: contracts/core/OptionsPremiumPricer.sol [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+     * @param t is the time to expiration in years
+     * @param isPut is whether the option is a put option
+     */
+    function getPremium(
+        uint256 st,
+        uint256 t,
+        bool isPut
+    ) external returns (uint256 premium) {
+        uint256 sp = priceOracle.latestAnswer().div(priceOracle.decimals());
+        uint256 v = volatilityOracle.vol();
```
Comment (kenchangh) [inline contracts/core/OptionsPremiumPricer.sol:50]: We need to check if the the calculations use the volatility per period (`volatilityOracle.vol`) or annualized volatility (`volatilityOracle.annualizedVol`)

File: contracts/core/OptionsPremiumPricer.sol (hunk 1 above)
Comment (kenchangh) [reply contracts/core/OptionsPremiumPricer.sol:51]: We can assume 0% rate for our own purposes since we're doing weekly options. We can revisit this in the future when we do longer dated options. So we can just remove out adding the interest component tbh

**Comment by kenchangh** [comment]
Change looks good to me 👍  minor feedback

File: contracts/core/OptionsPremiumPricer.sol (hunk 5 above)
Comment (chuddster) [reply contracts/core/OptionsPremiumPricer.sol:50]: in [wikipedia](https://en.wikipedia.org/wiki/Black%E2%80%93Scholes_model) for black scholes there is no mention that volatility is annualized. It explicitly states that risk free rate and drift rate is annualized, so dont think volatility is annualized. Also, volatility in this case is the _standard deviation of log returns_. VolOracle do esn't do log returns right?

Comment (kenchangh) [reply contracts/core/OptionsPremiumPricer.sol:55]: I'm actually not sure if the volatility input passed should be annualized, weekly or over 6 hours (what `vol` returns now). Will need to test this out.

>VolOracle do esn't do log returns right?

Yea the VolOracle just takes the most basic definition of volatility which is the standard deviation of returns. There are a few conflicting definitions out there that involve using the natural log of returns

Comment (chuddster) [reply contracts/core/OptionsPremiumPricer.sol:55]: do you need anything from me here?

Comment (kenchangh) [reply contracts/core/OptionsPremiumPricer.sol:55]: Nope, we can proceed with what we have here first

File: contracts/core/OptionsPremiumPricer.sol [hunk 6]
```diff
@@ -18,18 +18,23 @@ contract OptionsPremiumPricer {
     IVolatilityOracle public immutable volatilityOracle;
     IPriceOracle public immutable priceOracle;
     ICompToken public immutable cToken;
+    // IKEEP3rVolatility: 0xCCdfCB72753CfD55C5afF5d98eA5f9C43be9659d
+    IBlackScholes public immutable blackScholes;
```
Comment (kenchangh) [inline contracts/core/OptionsPremiumPricer.sol:22]: Would you want to just make this a library?

Comment (kenchangh) [reply contracts/core/OptionsPremiumPricer.sol:22]: Seems cleaner since it's just an immutable variable anyway

Comment (chuddster) [reply contracts/core/OptionsPremiumPricer.sol:22]: you're referring to IBlackScholes? this is the [IKEEP3rVolatility](0xCCdfCB72753CfD55C5afF5d98eA5f9C43be9659d) contract. Do you mean copy pasting the quoteAll method from the contract and making it an internal library instead?

File: contracts/core/OptionsPremiumPricer.sol [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        uint256 expiryTimestamp,
+        bool isPut
+    ) external view returns (uint256 premium) {
+        require(
+            expiryTimestamp > block.timestamp,
+            "Expiry must be in the future!"
+        );
+
+        uint256 sp = priceOracle.latestAnswer().div(10**priceOracle.decimals());
+        uint256 v = volatilityOracle.vol();
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:62]: blackScholes.quotePrice(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2, 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48, t, sp, st);

If we want to use vol from IKeeprV3 instead of homemade vol

File: contracts/core/OptionsPremiumPricer.sol [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        uint256 t = expiryTimestamp.sub(block.timestamp).div(1 days);
+
+        console.log("strike price is %s", st);
+        console.log("underlying price is %s", sp);
+        console.log("volatility is %s", v);
+        console.log("time to expiry in days is %s", t);
+
+        (uint256 call, uint256 put) = blackScholes.quoteAll(t, v, sp, st);
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:70]: some fundamental problems with blackScholes.quoteAll.

ex: if we input 
      a. t = 7 (it should be in days because [quoteAll](https://github.com/keep3r-network/keep3r.network/blob/a6897007db6e656e0e310ed7ee4ad42904fe2794/contracts/Keep3rV1Volatility.sol#L193) subsequently divides by 365.
      b. vol = 44478303 (disregard that this is wrong)
      c. sp (underlying) = 2000
      d. st (strike) = 2200

The call option costs more than the put option even though the call option would be OTM and the put would be ITM when purchased.

Other issues include call options with lower strike prices are **cheaper** than higher strike prices, expiries further out are cheaper. 

Bottom line not reliable, will have to change

File: contracts/libraries/Math.sol [hunk 9]
```diff
@@ -78,4 +95,190 @@ library Math {
...
+     * input range: 0 <= x <= OPT_EXP_MAX_VAL - 1
+     * auto-generated via 'PrintFunctionOptimalExp.py'
+     * Detailed description:
+     * - Rewrite the input as a sum of binary exponents and a single residual r, as small as possible
+     * - The exponentiation of each binary exponent is given (pre-calculated)
+     * - The exponentiation of r is calculated via Taylor series for e^x, where x = r
+     * - The exponentiation of the input is calculated by multiplying the intermediate results above
+     * - For example: e^5.521692859 = e^(4 + 1 + 0.5 + 0.021692859) = e^4 * e^1 * e^0.5 * e^0.021692859
+     */
+    function optimalExp(uint256 x) internal pure returns (uint256) {
```
Comment (chuddster) [inline contracts/libraries/Math.sol:110]: you think its worth adding tests for these additional methods?

File: contracts/core/OptionsPremiumPricer.sol [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+        premium = !isPut ? call : put;
+    }
+
+    /**
+     * @notice Calculates the risk free interest rate on the premium underlying asset
+     * using Compounds's risk free interest rate
+     * Formula reference: https://compound.finance/docs (Calculating the APY Using Rate Per Block)
+     */
+    function getAssetRiskFreeRate() private view returns (uint256 apy) {
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:82]: getAssetRiskFreeRate entails some tricky math for getting decimals right - its not correct yet. will either work out it out if we need to use or delete it

File: contracts/core/OptionsPremiumPricer.sol [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * @notice Calculates the option's delta
+     * Formula reference: `d_1` in https://www.investopedia.com/terms/b/blackscholes.asp
+     * http://www.optiontradingpedia.com/options_delta.htm
+     * https://www.macroption.com/black-scholes-formula/
+     * @param st is the strike price of the option
+     * @param expiryTimestamp is the unix timestamp of expiry
+     */
+    function getOptionDelta(uint256 st, uint256 expiryTimestamp)
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:113]: also janky - will need to update

File: contracts/core/OptionsPremiumPricer.sol (hunk 6 above)
Comment (kenchangh) [reply contracts/core/OptionsPremiumPricer.sol:24]: Oh I see- I didn't know we were using an external contract for the black scholes calculation. I'm all for using whatever keeper has.

File: contracts/core/OptionsPremiumPricer.sol (hunk 10 above)
Comment (kenchangh) [reply contracts/core/OptionsPremiumPricer.sol:79]: We should just delete it imo

File: contracts/core/OptionsPremiumPricer.sol (hunk 8 above)
Comment (kenchangh) [reply contracts/core/OptionsPremiumPricer.sol:67]: >The call option costs more than the put option even though the call option would be OTM and the put would be ITM when purchased.

>Other issues include call options with lower strike prices are cheaper than higher strike prices, expiries further out are cheaper.

Do we know the underlying reasons for why quoteAll is unreliable? They must have made mistakes in how they compute BS

File: contracts/core/OptionsPremiumPricer.sol [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+    /**
+     * @notice Calculates black scholes for both put and call
+     * @param t is the days until expiry
+     * @param v is the annualized volatility
+     * @param sp is the underlying price
+     * @param st is the strike price
+     * @return call is the premium of the call option given parameters
+     * @return put is the premium of the put option given parameters
+     */
+    function quoteAll(
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:117]: if you compare this to original keep3rvolatility [quoteAll](https://github.com/keep3r-network/keep3r.network/blob/a6897007db6e656e0e310ed7ee4ad42904fe2794/contracts/Keep3rV1Volatility.sol#L172), the only thing I changed was in st > sp condition, instead of _c = st-sp+_p we do _c = sp-st+_p.

their solution did not make sense because st > sp means that strike price is greater than underlying, which means the call should be less expensive than the put, which means that _c < _p so it should be (sp - st) which is negative as opposed to (st - sp) which is positive.

File: contracts/core/OptionsPremiumPricer.sol [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+            uint256 t
+        )
+    {
+        // chainlink oracle returns crypto / usd pairs with 8 decimals, like otoken strike price
+        sp = priceOracle.latestAnswer().div(10**priceOracle.decimals()).mul(
+            10**8
+        );
+        // annualized vol * 10 ** 7 because delta expects 16 decimals
+        // and annualizedVol is 8 decimals + it shifts percentage by one decimal place
+        v = volatilityOracle.annualizedVol().mul(10**7);
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:217]: there is some decimal weirdness wrt volatility. apparently shifting it to only 16 decimals aligns it with online calculators (ex: 1.2% vol actually resolves to 120% vol so I just multiplied by 10 ** 7 instead of 10 ** 9)

good resources to double check my black scholes:

https://goodcalculators.com/black-scholes-calculator/
https://www.calkoo.com/en/black-scholes-option-pricing-model

File: contracts/core/OptionsPremiumPricer.sol [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+        (uint256 sp, uint256 v, uint256 t) =
+            blackScholesParams(expiryTimestamp);
+
+        uint256 d1;
+        uint256 d2;
+
+        // Divide delta by 10 ** 12 to bring it to 2 decimals for strike selection
+        if (sp >= st) {
+            (d1, d2) = d(t, v, sp, st);
+            delta = Math.ncdf((Math.FIXED_1 * d1) / 1e18).div(10**12);
```
Comment (chuddster) [inline contracts/core/OptionsPremiumPricer.sol:97]: dividing by 10 ** 12 here because strike selection module expects 2 decimals ( ex: delta 1 is 100, delta 0.89 is 89)

File: contracts/core/OptionsPremiumPricer.sol [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+
+contract OptionsPremiumPricer is DSMath {
+    using SafeMath for uint256;
+
+    /**
+     * Immutables
+     */
+    IVolatilityOracle public immutable volatilityOracle;
+    IPriceOracle public immutable priceOracle;
+    ICompToken public immutable cToken;
```
Comment (kenchangh) [inline contracts/core/OptionsPremiumPricer.sol:20]: We can remove ctoken references since we're not using them anymore

File: contracts/interfaces/ICompToken.sol [hunk 16]
```diff
@@ -0,0 +1,6 @@
+//SPDX-License-Identifier: GPL-3.0
```
Comment (kenchangh) [inline contracts/interfaces/ICompToken.sol:1]: Can remove this file

File: test/core/OptionsPremiumPricer.ts [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+  });
+
+  describe.skip("getAssetRiskFreeRate", () => {
+    time.revertToSnapshotAfterEach();
+
+    it("gets reasonable APY (~0%-0.75%)", async function () {
+      const apy = await optionsPremiumPricer.getAssetRiskFreeRate();
+      console.log(apy);
+    });
+  });
```
Comment (kenchangh) [inline test/core/OptionsPremiumPricer.ts:56]: Can remove this reference

File: test/core/OptionsPremiumPricer.ts [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+      const expiryTimestamp = (await time.now()).add(WEEK);
+      const isPut = true;
+
+      const premium = await optionsPremiumPricer.getPremium(
+        strikePrice,
+        expiryTimestamp,
+        isPut
+      );
+
+      assert.isAbove(
```
Comment (kenchangh) [inline test/core/OptionsPremiumPricer.ts:101]: Would be helpful to include `assert.equal` tests so we know what the outputs are

File: test/core/OptionsPremiumPricer.ts [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+        .mul(BigNumber.from(10).pow(8));
+      const expiryTimestamp = (await time.now()).add(WEEK);
+
+      const delta = await optionsPremiumPricer.getOptionDelta(
+        strikePrice,
+        expiryTimestamp
+      );
+
+      console.log(delta.toString());
+      assert.isBelow(parseInt(delta.toString()), 50);
```
Comment (kenchangh) [inline test/core/OptionsPremiumPricer.ts:272]: Would be helpful to test exact delta values

**Review by kenchangh — DISMISSED** [review]

**Review by kenchangh — APPROVED** [review]
